network-ntt/src/server.rs

```diff
@@ -0,0 +1,270 @@
+//!                                      |
+//!                                      +---- client: remote-sock1
+//! ```
+//!
+//! As NT state is shared all the services could provide a uniform
+//! access to the chain state. And one could build very flexible
+//! topology.
+use cbor_event;
+use futures::{future, prelude::*, stream::Stream, sync::mpsc};
+use network_core::server::{self, block::BlockService, transaction::TransactionService};
+use protocol::{Inbound, Message};
+use std::net::SocketAddr;
+use tokio::net::{TcpListener, TcpStream};
+
+/// Internal structure of network transport node.
+#[derive(Clone)]
+pub struct Server<N> {
+    node: N,
+}
+
+/// Sets up a listening TCP socket bound to the given address.
+/// If successful, returns an asynchronous stream of `TcpStream` socket.
+pub fn listen(
+    sockaddr: SocketAddr,
+) -> Result<impl Stream<Item = TcpStream, Error = tokio::io::Error>, tokio::io::Error> {
+    let listener = TcpListener::bind(&sockaddr)?;
+    let stream = listener.incoming();
+    Ok(stream)
+}
+
+/// Run a server that will listen on a specific sockets
+/// and accept all incomming connections.
+/// Server maintains all of the incomming connection and
+/// `run_connection` is spawned on each of those connections.
+pub fn accept<N: 'static, F>(
+    stream: TcpStream,
+    node: Server<N>,
+) -> impl future::Future<Item = impl futures::future::Future<Item=(),Error=()>, Error = ()>
+where
+    N: server::Node + Clone,
+    F: future::Future<Item = (), Error = ()>,
+    <<N as server::Node>::BlockService as server::block::BlockService>::Block : cbor_event::de::Deserialize,
+    <<N as server::Node>::BlockService as server::block::BlockService>::Block : cbor_event::se::Serialize,
+    <<N as server::Node>::BlockService as server::block::BlockService>::BlockId : cbor_event::de::Deserialize,
+    <<N as server::Node>::BlockService as server::block::BlockService>::BlockId : cbor_event::se::Serialize,
+    <<N as server::Node>::BlockService as server::block::BlockService>::Header : cbor_event::de::Deserialize,
+    <<N as server::Node>::BlockService as server::block::BlockService>::Header : cbor_event::se::Serialize,
+    <<N as server::Node>::TransactionService as server::transaction::TransactionService>::TransactionId : cbor_event::de::Deserialize,
+    <<N as server::Node>::TransactionService as server::transaction::TransactionService>::TransactionId : cbor_event::se::Serialize,
+{
+    protocol::Connection::accept(stream)
+        .map_err(|_| ())
+        .and_then(move |connection| {
+            let node = node.clone();
+            Ok(run_connection(node, connection))
+        })
+}
+
+/// Connect to another client.
+/// `run_connection` is spawned on the single heavyweight
+/// connection.
+pub fn connect<N: 'static, F>(
+    sockaddr: SocketAddr,
+    node: Server<N>,
+) -> impl future::Future<Item = impl futures::future::Future<Item=(),Error=()>, Error = ()>
```

Comment:
So, `accept` and `connect` return different anonymous `Future` types backed by the protocol connection. Does it mean that we have asymmetric protocol connections in NTT after all?